pkg/issuer/acme/http/solver/solver.go

```diff
@@ -0,0 +1,61 @@
+package solver
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"path"
+	"strings"
+)
+
+const (
+	CertManagerSelfTestParam = "selftest"
+)
+
+type HTTP01Solver struct {
+	ListenPort int
+
+	Domain string
+	Token  string
+	Key    string
+}
+
+func (h *HTTP01Solver) Listen() error {
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.EscapedPath() == "/" || r.URL.EscapedPath() == "/healthz" {
+			log.Printf("responding ok to health check")
+			w.WriteHeader(http.StatusOK)
+			return
+		}
+
+		// extract vars from the request
+		host := strings.Split(r.Host, ":")[0]
+		basePath := path.Dir(r.URL.EscapedPath())
+		token := path.Base(r.URL.EscapedPath())
+
+		log.Printf("got request for host %s, basePath %s, token %s", host, basePath, token)
+		// verify the base path is correct
+		if basePath != HTTPChallengePath {
+			log.Printf("invalid basePath - expected %s", HTTPChallengePath)
+			http.NotFound(w, r)
+			return
+		}
+
+		log.Printf("comparing host %s against %s", host, h.Domain)
+		// if either the host or the token don't match what is expected,
+		// we should continue to the next loop iteration
+		if h.Domain != host || h.Token != token {
+			// if nothing else, we return a 404 here
+			http.NotFound(w, r)
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprint(w, h.Key)
+
+		if r.URL.Query().Get(CertManagerSelfTestParam) == "" {
+			// os.Exit(0)
```

Comment:
So the issue I'm having here, is that the GCLB seems to make a request to the base path that is programmed in when creating a LB. This request does *not* contain the `selftest` parameter either, causing the job to exit early (before the selftest or the actual acme validation has happened).

If we were to use the actual base acme challenge path (ie. `/.well-known/acme-challenge`) instead of the full path including token in the ingress resource, I don't think this would be an issue as an invalid token would be provided to it, and we could just respond with 200 (assuming this is some kind of health check). However, if we do this, we run into issues with non-standard path wildcards between ingress controllers (ie. gclb requires a '*' whereas nginx-ingress does prefix based matching already).

I'm not sure how I'm best to overcome this. It seems the cleanest way (due to wildcards) is to include the token as part of the Ingress path, however doing so means I cannot safely call 'os.Exit'. I guess a short-term workaround would be to *only* call `os.Exit` *IF* the `selftest` param isn't set *and* we've previously had a request *with* the `selftest` param set?